### DISCUSSION TITLE
Running MinIO in OpenShift with enabled SELinux

### DISCUSSION THREAD

**Discussion by dnskr** [minio/minio#14854] [Q&A] [answered]
Hi!
I use minio-operator to deploy minio in Kubernetes and OpenShift clusters. During one of the minio upgrades in OpenShift cluster I found that pods cannot start again.
The reason is that SELinux is enabled and it relabeling files each time newly created pod mounts a persistent volume. This relabeling cannot be done within specific time period (cannot be increased) for volumes with huge amount of files. So pods continuously failing with "context timeline exceeded" error message.

Are any known recommendations or solutions regarding the problem with SELinux in MinIO setup?

**Comment by steve-gray** [comment]
I'm hitting this one too @dnskr  - did you ever get a solution? We're not on OphenShift though, but this will be impacting all cloud solutions equally if they use SELinux. Ours just hangs effectively forever with:

```
 Error: Kubelet may be retrying requests that are timing out in CRI-O due to system load: context deadline exceeded: error reserving ctr name k8s_minio-internal-ss-0-0_storage_a1e9d7a4-5718-4240-b713-b80f019347a2_0 for id e8ebfc0cc416396a070eba69b729e41788a4584e4e4b577bd0ad0b18b461e9ea: name is reserved
```

**Reply by dnskr** [reply]
> I'm hitting this one too @dnskr - did you ever get a solution? We're not on OphenShift though, but this will be impacting all cloud solutions equally if they use SELinux. Ours just hangs effectively forever with:
> 
> ```
>  Error: Kubelet may be retrying requests that are timing out in CRI-O due to system load: context deadline exceeded: error reserving ctr name k8s_minio-internal-ss-0-0_storage_a1e9d7a4-5718-4240-b713-b80f019347a2_0 for id e8ebfc0cc416396a070eba69b729e41788a4584e4e4b577bd0ad0b18b461e9ea: name is reserved
> ```

In non-OpenShift cluster you can increase runtime-request-timeout parameter value for kubelet on minio nodes:

1. Deprecated way https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet
```
--runtime-request-timeout duration     Default: 2m0s 
```
2. Recommended way https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1
```
runtimeRequestTimeout
[meta/v1.Duration](https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1#Duration)
```

But OpenShift 4.10 does not allow to specify `RuntimeRequestTimeout` parameter
https://docs.openshift.com/container-platform/4.10/nodes/nodes/nodes-nodes-managing.html

**Reply by steve-gray** [reply]
That's not really an answer as for real world high volume numbers of files this becomes a timeout war and any cluster operations end up leading easily to long downtimes as you wait for pods to come back online.

We can't just hike the timeouts into infinity.

**Reply by dnskr** [reply]
You are right. That's why I didn't post it as the answer :) This is more a workaround to make it work in the short term.

**Comment by dnskr** [comment, marked as answer]
Possible solution would be to use `io.kubernetes.cri-o.TrySkipVolumeSELinuxLabel` annotation.
See https://github.com/cri-o/cri-o/blob/main/pkg/annotations/annotations.go
```
// TrySkipVolumeSELinuxLabelAnnotation is the annotation used for optionally skipping relabeling a volume
// with the specified SELinux label.  The relabeling will be skipped if the top layer is already labeled correctly.
TrySkipVolumeSELinuxLabelAnnotation = "io.kubernetes.cri-o.TrySkipVolumeSELinuxLabel"
```

**The following steps need to be done:**
1. Add CRI-O configuration file /etc/crio/crio.conf.d/01-selinux.conf
```
[crio.runtime.runtimes.selinux]
runtime_path = "/usr/bin/runc"
runtime_root = "/run/runc"
runtime_type = "oci"
allowed_annotations = ["io.kubernetes.cri-o.TrySkipVolumeSELinuxLabel"]
```
2. Add RuntimeClass to the cluster
```
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: selinux
handler: selinux
```
3. Create MinIO pods with specific annotation and runtimeClassName.
```
apiVersion: v1
kind: Pod
metadata:
  ...
  annotations:
    io.kubernetes.cri-o.TrySkipVolumeSELinuxLabel: "true"
spec:
  runtimeClassName: selinux
```
4. Add `runtimeClassName` support to
     - [MinIO Operator](https://github.com/minio/operator)
     - [MinIO helm chart](https://github.com/minio/minio/tree/master/helm/minio)


FYI @steve-gray

**Reply by harshavardhana** [reply]
@dnskr feel free to open an issue on this in the relevant repo.

**Reply by dnskr** [reply]
The issues have been created:
- Operator repo - https://github.com/minio/operator/issues/1194
- MinIO repo - https://github.com/minio/minio/issues/15217
